### DISCUSSION TITLE
Announcing Ariadne GraphQL Modules

### DISCUSSION THREAD

**Discussion by rafalp** [mirumee/ariadne#841] [Announcements]
I am happy to announce that [Ariadne GraphQL Modules](https://github.com/mirumee/ariadne-graphql-modules) library has just been released 🎉

This library provides new way to define your GraphQL schema and organize your python code:

```python
from datetime import date

from ariadne.asgi import GraphQL
from ariadne_graphql_modules import ObjectType, gql, make_executable_schema


class Query(ObjectType):
    __schema__ = gql(
        """
        type Query {
            message: String!
            year: Int!
        }
        """
    )

    @staticmethod
    def resolve_message(*_):
        return "Hello world!"

    @staticmethod
    def resolve_year(*_):
        return date.today().year


schema = make_executable_schema(Query)
app = GraphQL(schema=schema, debug=True)
```

Please see [README](https://github.com/mirumee/ariadne-graphql-modules/blob/main/README.md) for examples and [REFERENCE](https://github.com/mirumee/ariadne-graphql-modules/blob/main/REFERENCE.md) for early documentation.

Remember that that this is first general release of Ariadne GraphQL Modules. We are pretty sure breaking changes will happen but we feel that library is complete enough to let you build complete GraphQL schema and we'll love to hear what you think about it and what would you like to see in future releases :)

**Comment by betaboon** [comment]
would it be possible to combine this somehow to loading schemas from path?
i like having the schemas in seperate `.graphql`-files :)

**Reply by rafalp** [reply]
Nope. Modules put together type's GraphQL definition and python logic and let you spread your schema in multiple python classes, that's the selling point for them.

**Reply by betaboon** [reply]
ah ok. thanks for clearing that up.
what i am actually looking for is being able to define the schema in `.graphql`-files but have all the resolvers in class-structures

**Reply by TabishB** [reply]
@betaboon 
While maybe against the philosophy of the package, it is technically possible by using `load_schema_from_path`
```
from ariadne import load_schema_from_path
from ariadne_graphql_modules import ObjectType

class UserType(ObjectType):
    __schema__ = load_schema_from_path("/path/to/schema/user.graphql")
```

**Comment by bartenra** [comment]
Just tried it out for the first time, but I'm getting `type object 'Query' has no attribute 'bind_to_schema'`...

from ariadne_graphql_modules import ObjectType, gql, InputType, EnumType, DeferredType, convert_case

```py
from ariadne_graphql_modules import ObjectType, gql, InputType, EnumType, DeferredType, convert_case

class Query(ObjectType):
    __schema__ = gql(
        '''
type Query {
  appointmentAvailabilitySet(slots: [AppointmentSlotInput!]!): [AppointmentSlot!]!
}
        '''
    )

    __requires__ = [AppointmentSlotType, AppointmentSlotInputType]

    __aliases__ = convert_case

    @staticmethod
    def resolve_appointment_availability_set(*_, slots):
        print(slots)
        pass
```

Passing it to `make_executable_schema`.

**Reply by rafalp** [reply]
Note that `ariadne_graphql_modules` exports it's own `make_executable_schema`, and you can't mix approaches from those two libraries with each other.

**Reply by bartenra** [reply]
Ah makes sense, thank you! Bummer, that makes gradual adoption not an option.

**Reply by rafalp** [reply]
It doesn't, but making a jump over to try it is not a time consuming effort. It took me ~6h to move my project from Ariadne to GraphQL Modules, and majority of that time was copying SDL and resolvers code from old implementation to new.

**Comment by rafalp** [comment]
@bartenra starting with Ariadne GraphQL Modules 0.7 it's `ariadne_graphql_modules` now works with both new and old schema definitions:

https://github.com/mirumee/ariadne-graphql-modules/blob/main/MOVING.md
